sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -76,8 +76,36 @@ case class SortMergeJoinExec(
         s"${getClass.getSimpleName} should not take $x as the JoinType")
   }
 
-  override def requiredChildDistribution: Seq[Distribution] =
-    HashClusteredDistribution(leftKeys) :: HashClusteredDistribution(rightKeys) :: Nil
+  private def avoidShuffleIfPossible(
+      joinKeys: Seq[Expression],
+      expressions: Seq[Expression]): Seq[Distribution] = {
+    val indices = expressions.map(x => joinKeys.indexWhere(_.semanticEquals(x)))
```

Comment:
what if here we don't find an expression? I think it would return -1 causing an error when using the index later. Can we also add a test case for this situation?

Reply:
```      
      case HashPartitioning(leftExpressions, _)
        if leftPartitioning.satisfies(ClusteredDistribution(leftKeys)) =>
        avoidShuffleIfPossible(leftKeys, leftExpressions)
```
` if leftPartitioning.satisfies(ClusteredDistribution(leftKeys))` has ensured `expressions` is a subset of `joinKeys`, so it would not return -1, right?

Reply:
yes, you're right, thanks.